### PR TITLE
Rest api integration

### PR DESCRIPTION
**Pull request by shogunpurple**
## Description

Addresses https://github.com/Budibase/budibase/issues/1024

This PR contains code allowing budibase users to interact with REST APIs as an external datasource.

Each CRUD method maps to a particular HTTP verb:
**Create** - POST
**Read** - GET
**Update** - PUT
**Delete** - DELETE

You can pass request headers and bodies with the relevant requests, and these can all be parameterised with values from the frontend. See the screencasts below to see different uses of the new integration interacting with a mock data API.

### Usage
GET
https://www.loom.com/share/66f6aabde8184d9fbccb97ccfc359bb1

POST
https://www.loom.com/share/afb129e5f8a447709a4d026ba4a597fa

### CODE REVIEW

File: packages/server/src/integrations/rest.js [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+      method: "POST",
+      headers,
+      body: JSON.stringify(json),
+    })
+
+    return await response.json()
+  }
+
+  async delete({ path, headers = {} }) {
+    const response = await fetch(this.config.url + path, {
```
Comment (mike12345567) [inline packages/server/src/integrations/rest.js:104]: Might be a good idea to have a function to build the URL and in this function check for any `//`, e.g. the user added an extra slash on the end of the base URL, or one on the start etc.

File: packages/server/src/integrations/rest.js [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+class RestIntegration {
+  constructor(config) {
+    this.config = config
+  }
+
+  async create({ path, headers = {}, json }) {
+    const response = await fetch(this.config.url + path, {
+      method: "POST",
+      headers,
+      body: JSON.stringify(json),
```
Comment (mike12345567) [inline packages/server/src/integrations/rest.js:79]: Currently this only allows use of JSON based API, but in the videos you showed adding JSON headers like `Content-Type` - either might be good to hard-code JSON headers and make them non-editable so its obvious that it currently only supports JSON, or use the `Accept` header to work out whether to parse as JSON, XML, plaintext etc

**Comment by mike12345567** [comment]
Looks good to me bar the couple of comments, one thing I would say is that I think it'd be great to change over the "functions" from create, read, update, delete to the real HTTP methods, I think in its current form its a little confusing as I would think of a create as a PUT and an update as a PATCH, but some people would use a POST for a create etc. I think currently the system expects the `read` function to exist is a couple of places, but if we fixed this then the `GET` method could have the `readable: true` property in the schema and the rest could just be custom ones of `POST`, `DELETE`, `PATCH`, `PUT` etc.

**Comment by shogunpurple** [comment]
Cheers @mike12345567 - Joe showed me some great designs you worked on together, I'm going to update this feature to use the new designs and to name the methods after the HTTP verbs rather than CRUD - I think that makes sense.

File: packages/server/src/integrations/rest.js (hunk 2 above)
Comment (shogunpurple) [reply packages/server/src/integrations/rest.js:79]: Good point - I considered adding an option to choose the "type" of the body/response, but at the moment it doesn't make a whole lot of sense to return anything other than JSON based on the fact that the schema gets built from an array of objects. 

I'll add some parsing based on the accept header so there's a bit more flexibility - that makes sense 👍

**Review by aptkingston — APPROVED** [review]
LGTM! This also technically allows someone to link up any other datasource they want by writing their own API middleware to interact with the datasource, so opens loads of possibilities for developers or datasources which need things like custom auth.

**Pull request merged**